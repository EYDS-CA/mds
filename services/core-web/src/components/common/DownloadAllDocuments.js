import React from "react";
import { notification } from "antd";
import PropTypes from "prop-types";
import { getDocumentDownloadToken } from "@common/utils/actionlessNetworkCalls";
import { DownloadOutlined } from "@ant-design/icons";

const propTypes = {
  submissions: PropTypes.arrayOf(PropTypes.any).isRequired,
};

const downloadDocument = (url) => {
  const a = document.createElement("a");
  a.href = url.url;
  a.download = url.filename;
  a.style.display = "none";
  document.body.append(a);
  a.click();
  a.remove();
};

const waitFor = (conditionFunction) => {
  const poll = (resolve) => {
    if (conditionFunction()) resolve();
    else setTimeout(() => poll(resolve), 400);
  };

  return new Promise(poll);
};

export const DownloadAllDocuments = (props) => {
  const handleDownloadAll = () => {
    const docURLS = [];

    const totalFiles = props.submissions.length;
    if (totalFiles === 0) {
      return;
    }

    props.submissions.forEach((doc) =>
      getDocumentDownloadToken(doc.documentManagerGuid, doc.filename, docURLS)
    );

    let currentFile = 0;
    waitFor(() => docURLS.length === props.submissions.length).then(async () => {
      // eslint-disable-next-line no-restricted-syntax
      for (const url of docURLS) {
        currentFile += 1;

        downloadDocument(url);

        // eslint-disable-next-line
        await new Promise((resolve) => setTimeout(resolve, 3000));
      }
      notification.success({
        message: `Successfully Downloaded: ${totalFiles} files.`,
        duration: 10,
      });
    });
  };
  return (
    <div className="custom-menu-item">
      <button type="button" className="full" onClick={() => handleDownloadAll()}>
        <DownloadOutlined style={{ color: "#5e46a1", paddingLeft: "8px", paddingRight: "22px" }} />
        Download All
      </button>
    </div>
  );
};

export default DownloadAllDocuments;
