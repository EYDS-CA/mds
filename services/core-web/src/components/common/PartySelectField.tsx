import React, { FC, useCallback, useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { debounce } from "lodash";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faHashtag } from "@fortawesome/pro-light-svg-icons";

import { Divider } from "antd";
import { MailOutlined, PhoneOutlined, PlusOutlined } from "@ant-design/icons";
import { Field } from "redux-form";
import { getSearchResults } from "@mds/common/redux/selectors/searchSelectors";
import { getLastCreatedParty } from "@mds/common/redux/selectors/partiesSelectors";
import { fetchSearchResults } from "@mds/common/redux/actionCreators/searchActionCreator";
import { setAddPartyFormState } from "@mds/common/redux/actionCreators/partiesActionCreator";
import { createItemIdsArray, createItemMap } from "@common/utils/helpers";
import { Validate } from "@common/utils/Validate";
import LinkButton from "@mds/common/components/common/LinkButton";
import RenderSelect from "@mds/common/components/forms/RenderSelect";

const renderAddPartyFooter = (showAddParty, partyLabel) => (
  <div className="wrapped-text">
    <Divider style={{ margin: "0" }} />
    <p className="footer-text">{`Can't find the ${partyLabel} you are looking for?`}</p>
    <LinkButton onClick={showAddParty}>
      <PlusOutlined className="padding-small--right" />
      {`Add a new ${partyLabel}`}
    </LinkButton>
  </div>
);

const transformData = (data, options, header) => {
  if (data.length === 0) {
    return [];
  }
  const transformedData = data
    .map((opt) => ({
      value: options[opt].party_guid,
      originalValue: options[opt],
      label: (
        <div>
          <span>{options[opt].name}</span>
          <div className="inline-flex">
            <div className="padding-right">
              <FontAwesomeIcon icon={faHashtag} />
            </div>
            <span>{options[opt].party_orgbook_entity?.registration_id}</span>
          </div>
          <div className="inline-flex">
            <div className="padding-right">
              <MailOutlined className="icon-xs" />
            </div>
            <span>
              {Validate.checkEmail(options[opt].email) ? options[opt].email : "Email Unknown"}
            </span>
          </div>
          <div className="inline-flex">
            <div className="padding-right">
              <PhoneOutlined className="icon-xs" />
            </div>
            <span>
              {options[opt].phone_no} {options[opt].phone_ext ? `x${options[opt].phone_ext}` : ""}
            </span>
          </div>
        </div>
      ),
    }))
    .filter((opt) => opt.value);

  // Display header only if desired (Add new party behavior is enabled.)
  if (header) {
    transformedData.unshift({ value: "header", label: header });
  }

  return transformedData;
};

interface PartySelectFieldProps {
  id?: string;
  name?: string;
  label?: string;
  partyLabel?: string;
  // set to true to only see people
  person?: boolean;
  // set to true to only see organizations
  organization?: boolean;
  allowAddingParties: boolean;
  validate: any[];
  searchResults?: any;
  initialValues?: any;
  allowNull?: boolean;
  disabled?: boolean;
  onSelect?: any;
}

export const PartySelectField: FC<PartySelectFieldProps> = ({
  initialValues,
  onSelect,
  person,
  organization,
  partyLabel = "contact",
  allowAddingParties,
  validate,
  disabled,
  allowNull,
  ...rest
}) => {
  const dispatch = useDispatch();

  const searchResults = useSelector(getSearchResults);
  const lastCreatedParty = useSelector(getLastCreatedParty);

  const [selectedOption, setSelectedOption] = useState({ value: "", label: "" });
  const [partyDataSource, setPartyDataSource] = useState([]);
  const [searching, setSearching] = useState(false);

  const handleFetchSearchResults = useCallback(
    (searchTerm: string, searchType: string) => {
      setSearching(true);
      dispatch(fetchSearchResults(searchTerm, searchType));
    },
    [dispatch]
  );

  const debouncedSearch = useMemo(
    () => debounce(handleFetchSearchResults, 1000),
    [handleFetchSearchResults]
  );

  useEffect(() => {
    if (searchResults) {
      setSearching(false);
    }
  }, [searchResults]);

  useEffect(() => {
    if (initialValues?.label) {
      handleSearch(initialValues.label);
      setSelectedOption({
        value: initialValues.value,
        label: initialValues.label,
      });
    }
  }, [initialValues]);

  const showAddPartyForm = () => {
    dispatch(
      setAddPartyFormState({
        showingAddPartyForm: true,
        person: !person,
        organization: !organization,
        partyLabel: partyLabel,
        initialValues,
      })
    );
  };

  useEffect(() => {
    if (searchResults || lastCreatedParty) {
      let filteredParties = searchResults?.party?.map((sr) => sr.result);

      if (filteredParties) {
        if (organization && !person) {
          filteredParties = filteredParties.filter(
            ({ party_type_code }) => party_type_code === "ORG"
          );
        } else if (person && !organization) {
          filteredParties = filteredParties.filter(
            ({ party_type_code }) => party_type_code === "PER"
          );
        }

        if (lastCreatedParty) {
          filteredParties.unshift(lastCreatedParty);
        }

        const newPartyDataSource = transformData(
          createItemIdsArray(filteredParties, "party_guid"),
          createItemMap(filteredParties, "party_guid"),
          allowAddingParties && renderAddPartyFooter(showAddPartyForm, partyLabel)
        );

        if (newPartyDataSource) {
          setPartyDataSource(newPartyDataSource);
        }
      }
    }

    if (lastCreatedParty) {
      setSelectedOption({
        value: lastCreatedParty.party_guid,
        label: lastCreatedParty.name,
      });
    }
  }, [lastCreatedParty, searchResults, organization, person, allowAddingParties, partyLabel]);

  const handleSearch = (value) => {
    if (value.length >= 2) {
      setSearching(true);
      debouncedSearch(value, "party");
    }
    setSelectedOption({ value, label: value });
  };

  const handleSelect = (value, option) => {
    setSelectedOption(option);
    if (onSelect) {
      onSelect({
        ...option,
        ...value.originalValue,
      });
    }
  };

  // Validator to ensure the selected option is in the collection of available options.
  // This validator is appened to any validators passed in from the form in the render function below.
  // eslint-disable-next-line consistent-return
  const validOption = (value) => {
    // ignore this validation if an initialValue is passed in
    if (initialValues && initialValues !== selectedOption && !allowNull) {
      return partyDataSource.find((opt) => opt.value === value)
        ? undefined
        : `Invalid ${partyLabel}`;
    }
  };

  return (
    <Field
      className="left"
      loading={searching}
      disabled={disabled}
      {...rest}
      component={RenderSelect}
      handleSearch={handleSearch}
      handleSelect={handleSelect}
      validate={validate.concat(validOption)}
      dataSource={partyDataSource}
      selectedOption={selectedOption}
    />
  );
};

export default PartySelectField;
