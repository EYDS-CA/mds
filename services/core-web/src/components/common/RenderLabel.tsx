import React, { FC } from "react";
import { Typography, Divider, Row, Col, Form } from "antd";
import { BaseInputProps } from "@mds/common/components/forms/BaseInput";


interface RenderLabelProps extends BaseInputProps {
  indentText?: string;
  className?: string;
}
/**
 * 
 * @deprecated - only used in GenerateDocumentFormField, don't add to new cases
 */
const RenderLabel: FC<RenderLabelProps> = ({ className = "", ...props }) => {
  return (
    <Form.Item>
      <Row className={className}>
        {props.indentText && (
          <>
            <Col span={2}>
              <Typography.Paragraph disabled>{props.indentText}</Typography.Paragraph>
            </Col>
            <Col span={1}>
              <Divider type="vertical" plain />
            </Col>
          </>
        )}
        <Col span={props.indentText ? 21 : 24}>
          <Typography.Paragraph
            ellipsis={{
              expandable: true,
              rows: 3,
            }}
          >
            {props.input.value}
          </Typography.Paragraph>
        </Col>
      </Row>
    </Form.Item>
  );
};

export default RenderLabel;
