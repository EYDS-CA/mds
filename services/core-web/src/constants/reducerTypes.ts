// Documents
export const DOCUMENTS = "DOCUMENTS";

// Alerts
export const MINE_ALERTS = "MINE_ALERTS";