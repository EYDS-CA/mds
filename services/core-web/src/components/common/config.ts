import RenderAutoComplete from "@mds/common/components/forms/RenderAutoComplete";
import RenderAutoSizeField from "@mds/common/components/forms/RenderAutoSizeField";
import RenderCheckbox from "@mds/common/components/forms/RenderCheckbox";
import RenderDate from "@mds/common/components/forms/RenderDate";
import RenderField from "@mds/common/components/forms/RenderField";
import RenderFileUpload from "@mds/common/components/forms/RenderFileUpload";
import RenderGroupCheckbox from "@mds/common/components/forms/RenderGroupCheckbox";
import RenderMultiSelect from "@mds/common/components/forms/RenderMultiSelect";
import RenderRadioButtons from "@mds/common/components/forms/RenderRadioButtons";
import RenderSelect from "@mds/common/components/forms/RenderSelect";
import RenderCascader from "@mds/common/components/forms/RenderCascader";
import RenderGroupedSelect from "@mds/common/components/forms/RenderGroupedSelect";
import RenderLargeSelect from "@mds/common/components/forms/RenderLargeSelect";
import RenderTime from "@mds/common/components/forms/RenderTime";
import RenderLabel from "./RenderLabel";
import RenderYear from "@mds/common/components/forms/RenderYear";

export const renderConfig = {
  CHECKBOX: RenderCheckbox,
  GROUP_CHECK_BOX: RenderGroupCheckbox,
  AUTOCOMPLETE: RenderAutoComplete,
  AUTO_SIZE_FIELD: RenderAutoSizeField,
  CASCADER: RenderCascader,
  DATE: RenderDate,
  TIME: RenderTime,
  YEAR: RenderYear,
  FIELD: RenderField,
  SCROLL_FIELD: RenderAutoSizeField,
  SELECT: RenderSelect,
  LARGE_SELECT: RenderLargeSelect,
  MULTI_SELECT: RenderMultiSelect,
  RADIO: RenderRadioButtons,
  GROUPED_SELECT: RenderGroupedSelect,
  LABEL: RenderLabel,
  FILE_UPLOAD: RenderFileUpload,
};
