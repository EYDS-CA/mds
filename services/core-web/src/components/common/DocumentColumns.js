import React from "react";
import * as Strings from "@common/constants/strings";
import { Button, Popconfirm } from "antd";
import { TRASHCAN } from "@/constants/assets";
import {
  renderDateColumn,
  renderDocumentLinkColumn,
  renderTextColumn,
} from "./CoreTableCommonColumns";

export const documentNameColumn = (
  documentNameColumnIndex = "document_name",
  title = "File Name"
) => {
  return renderDocumentLinkColumn(documentNameColumnIndex, title, true);
};

export const uploadDateColumn = (uploadDateIndex = "upload_date", title = "Uploaded") => {
  return renderDateColumn(uploadDateIndex, title, true, null, Strings.EMPTY_FIELD);
};

export const uploadedByColumn = (uploadedByIndex = "update_user", title = "Uploaded By") => {
  return renderTextColumn(uploadedByIndex, title);
};

export const removeFunctionColumn = (
  removeFunction,
  hideColumn = false,
  documentNameColumnIndex = "document_name",
  documentParent = ""
) => {
  const className = hideColumn ? "column-hide" : "test";
  return {
    key: "remove",
    className,
    render: (record) => (
      <div align="right" className>
        <Popconfirm
          placement="topLeft"
          title={`Are you sure you want to delete ${record[documentNameColumnIndex]}?`}
          onConfirm={(event) => removeFunction(event, record.key, documentParent)}
          okText="Delete"
          cancelText="Cancel"
        >
          <Button ghost type="primary" size="small">
            <img src={TRASHCAN} alt="remove" style={{ width: "24px" }} />
          </Button>
        </Popconfirm>
      </div>
    ),
  };
};
