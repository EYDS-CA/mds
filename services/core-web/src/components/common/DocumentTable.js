import React from "react";
import PropTypes from "prop-types";
import { Table, Popconfirm, Button } from "antd";
import { formatDate, dateSorter, nullableStringSorter } from "@common/utils/helpers";
import { some } from "lodash";
import DocumentLink from "@/components/common/DocumentLink";
import { TRASHCAN } from "@/constants/assets";
import CustomPropTypes from "@/customPropTypes";
import { closeModal, openModal } from "@common/actions/modalActions";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { modalConfig } from "@/components/modalContent/config";

const propTypes = {
  documents: PropTypes.arrayOf(CustomPropTypes.documentRecord),
  isViewOnly: PropTypes.bool,
  // eslint-disable-next-line react/no-unused-prop-types
  removeDocument: PropTypes.func,
  archiveDocuments: PropTypes.func,
  excludedColumnKeys: PropTypes.arrayOf(PropTypes.string),
  additionalColumnProps: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      colProps: PropTypes.objectOf(PropTypes.string),
    })
  ),
  // eslint-disable-next-line react/no-unused-prop-types
  documentParent: PropTypes.string,
  documentColumns: PropTypes.arrayOf(PropTypes.object),
  openModal: PropTypes.func.isRequired,
  view: PropTypes.string.isRequired,
};

const defaultProps = {
  documents: [],
  isViewOnly: false,
  removeDocument: () => {},
  archiveDocuments: () => {},
  excludedColumnKeys: [],
  additionalColumnProps: [],
  documentColumns: null,
  documentParent: null,
  view: "standard",
};

const openArchiveModal = (event, props, documents) => {
  event.preventDefault();

  props.openModal({
    props: {
      title: `Archive ${props.documents?.length > 1 ? "Multiple Files" : "File"}`,
      closeModal: props.closeModal,
      handleSubmit: props.archiveDocuments,
      documents,
    },
    content: modalConfig.ARCHIVE_DOCUMENT,
    // width: "75vw",
  });
};

export const DocumentTable = (props) => {
  const isMinimalView = props.view === "minimal";

  let columns = [
    {
      title: "Name",
      key: "name",
      dataIndex: "name",
      sorter: !isMinimalView && nullableStringSorter("name"),
      render: (text, record) => {
        if (isMinimalView) {
          return record.name;
        }

        return (
          <div key={record.key} title="Name">
            <DocumentLink
              documentManagerGuid={record.document_manager_guid}
              documentName={record.name}
              truncateDocumentName={false}
            />
          </div>
        );
      },
    },
    {
      title: "Dated",
      key: "dated",
      dataIndex: "dated",
      sorter: !isMinimalView && dateSorter("dated"),
      defaultSortOrder: "descend",
      render: (text) => <div title="Dated">{formatDate(text)}</div>,
    },
    {
      title: "Category",
      key: "category",
      dataIndex: "category",
      sorter: !isMinimalView && nullableStringSorter("category"),
      render: (text) => <div title="Category">{text}</div>,
    },
    {
      title: "Uploaded",
      key: "uploaded",
      dataIndex: "uploaded",
      sorter: !isMinimalView && dateSorter("uploaded"),
      defaultSortOrder: "descend",
      render: (text) => <div title="Uploaded">{formatDate(text)}</div>,
    },
    {
      key: "remove",
      className: props.isViewOnly || !props.removeDocument ? "column-hide" : "",
      render: (text, record) => (
        <div className={props.isViewOnly || !props.removeDocument ? "column-hide" : ""}>
          <Popconfirm
            placement="topLeft"
            title={`Are you sure you want to delete ${record.name}?`}
            onConfirm={(event) => props.removeDocument(event, record.key, props?.documentParent)}
            okText="Delete"
            cancelText="Cancel"
          >
            <Button ghost type="primary" size="small">
              <img src={TRASHCAN} alt="Remove" />
            </Button>
          </Popconfirm>
        </div>
      ),
    },
    {
      key: "archive",
      className: props.isViewOnly || !props.archiveDocuments ? "column-hide" : "",
      render: (text, record) => (
        <div className={props.isViewOnly || !props.archiveDocuments ? "column-hide" : ""}>
          <Button
            ghost
            type="primary"
            size="small"
            onClick={(event) => openArchiveModal(event, props, [record])}
          >
            Archive
          </Button>
        </div>
      ),
    },
  ];

  if (isMinimalView) {
    props.excludedColumnKeys = props.excludedColumnKeys.concat(["remove", "archive", "category"]);
  }

  if (!some(props.documents, "dated")) {
    columns = columns.filter((column) => column.key !== "dated");
  }

  if (props?.excludedColumnKeys?.length > 0) {
    columns = columns.filter((column) => !props.excludedColumnKeys.includes(column.key));
  }

  if (props?.additionalColumnProps?.length > 0) {
    // eslint-disable-next-line no-unused-expressions
    props?.additionalColumnProps.forEach((addColumn) => {
      const columnIndex = columns.findIndex((column) => addColumn?.key === column.key);
      if (columnIndex >= 0) {
        columns[columnIndex] = { ...columns[columnIndex], ...addColumn?.colProps };
      }
    });
  }

  return (
    <Table
      pagination={false}
      columns={props?.documentColumns ?? columns}
      locale={{ emptyText: "No Data Yet" }}
      dataSource={props.documents}
      showHeader={!isMinimalView}
      size={isMinimalView ? "small" : undefined}
      rowClassName={isMinimalView ? "ant-table-row-minimal" : undefined}
    />
  );
};

DocumentTable.propTypes = propTypes;
DocumentTable.defaultProps = defaultProps;

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      openModal,
      closeModal,
    },
    dispatch
  );

export default connect(null, mapDispatchToProps)(DocumentTable);
