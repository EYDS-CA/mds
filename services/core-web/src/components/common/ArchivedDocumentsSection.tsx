import React from "react";
import DocumentTable from "./DocumentTable";
import { Typography } from "antd";
import { IMineDocument } from "@mds/common";
import { DeleteOutlined } from "@ant-design/icons";

interface ArchivedDocumentsSectionProps {
  documents: IMineDocument;
}

const ArchivedDocumentsSection = (props: ArchivedDocumentsSectionProps) => {
  const docs = props.documents.map((d) => {
    d.name = d.document_name;

    return d;
  });

  return (
    <div id="archived-documents">
      <Typography.Title level={4}>
        <DeleteOutlined className="violet" />
        &nbsp;Archived Documents{" "}
      </Typography.Title>
      <Typography.Paragraph>
        These files are not reviewed as part of the submission.
      </Typography.Paragraph>
      <DocumentTable documents={docs} excludedColumnKeys={["archive", "remove"]}>
        {" "}
      </DocumentTable>
    </div>
  );
};

export default ArchivedDocumentsSection;
