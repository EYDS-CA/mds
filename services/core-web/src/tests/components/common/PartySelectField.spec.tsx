import React from "react";
import { PartySelectField } from "@/components/common/PartySelectField";
import { render } from "@testing-library/react";
import { ReduxWrapper } from "@mds/common/tests/utils/ReduxWrapper";
import { SEARCH, PARTIES } from "@mds/common/constants/reducerTypes";
import * as MOCK from "@mds/common/tests/mocks/dataMocks";
import FormWrapper from "@mds/common/components/forms/FormWrapper";

const initialState = {
  [SEARCH]: { searchValues: { label: "Mock Party", value: "29489218432" } },
  [PARTIES]: MOCK.PARTY.parties[0],
};

describe("PartySelectField", () => {
  it("renders properly", () => {
    const { container } = render(
      <ReduxWrapper initialState={initialState}>
        <FormWrapper name="test_party_select" isEditMode={false} onSubmit={jest.fn()}>
          <PartySelectField validate={[]} allowAddingParties={true} name="test" />
        </FormWrapper>
      </ReduxWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
