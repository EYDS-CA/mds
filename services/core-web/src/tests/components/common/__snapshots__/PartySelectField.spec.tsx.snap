// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartySelectField renders properly 1`] = `
<div>
  <form
    class="ant-form ant-form-vertical common-form common-form-test_party_select form-view"
    id="test_party_select"
  >
    <div
      class="view-item ant-form-item"
    >
      
      <div
        class="ant-typography view-item-value"
      >
        N/A
      </div>
    </div>
  </form>
</div>
`;
